model/model.go

```diff
@@ -317,12 +340,23 @@ func build(kubeClient *kubernetes.Clientset, routerDeployment *v1beta1ext.Deploy
 	return routerConfig, nil
 }
 
-func buildRouterConfig(routerDeployment *v1beta1.Deployment, platformCertSecret *v1.Secret, dhParamSecret *v1.Secret) (*RouterConfig, error) {
+func buildRouterConfig(routerDeployment *v1beta1.Deployment, configMap *v1.ConfigMap, platformCertSecret *v1.Secret, dhParamSecret *v1.Secret) (*RouterConfig, error) {
 	routerConfig := newRouterConfig()
-	err := modeler.MapToModel(routerDeployment.Annotations, "nginx", routerConfig)
+
+	if configMap.Data != nil {
+		err := configmapModeler.MapToModel(configMap.Data, "nginx", routerConfig)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// To maintain backwards compatibility with previous Deis releases and Kube < 1.2
+	// annotations on the router Deployment override the ConfigMap.
+	err := annotationModeler.MapToModel(routerDeployment.Annotations, "nginx", routerConfig)
```

Comment:
Alright, let's break down the options and pros/cons of each:

1. Configuration done through deployment annotations
  - Pros:
    - No changes, though we'll still need to add test coverage
    - Consistent configuration: you configure apps via annotations, you configure router via annotations
  - Cons:
    - To configure router, operator must edit deployment manifest - most of which is not relevant to the configurations he/she would be making
    - Annotation keys are long with the prefix of "router.deis.io/"

2. Deployment annotations override configurations within a ConfigMap
  - Pros:
    - Using configmaps is the "expected" way to configure an application within Kubernetes
    - Isolates the operator-configurable parts of the router from the rest of the surrounding templated yaml that currently exists in the deployment manifest
  - Cons:
    - Multiple sources of configuration make it harder for an operator to track down where a specific configuration is coming from
    - Because of restrictions on configmap keys, the keys of the configmap and the annotation cannot exactly match, making documentation difficult
      - This is slightly mitigated by focusing the main documentation on the configmap method and link off to a page which describes the annotations override behavior

3. Use the configmap if it exists, otherwise fallback to annotations
  - Pros:
    - Using configmaps is the "expected" way to configure an application within Kubernetes
    - Isolates the operator-configurable parts of the router from the rest of the surrounding templated yaml that currently exists in the deployment manifest
  - Cons:
    - Because of restrictions on configmap keys, the keys of the configmap and the annotation cannot exactly match, making documentation difficult
      - This is slightly mitigated by focusing the main documentation on the configmap method and link off to a page which describes the annotations

4. Only use configmap
  - Pros:
    - Simple
    - Using configmaps is the "expected" way to configure an application within Kubernetes
    - Isolates the operator-configurable parts of the router from the rest of the surrounding templated yaml that currently exists in the deployment manifest
  - Cons:
    - Not compatible with Kube < 1.2
    - Not backwards-compatible with previous versions of Workflow
      - Script to migrate previous annotation-configs to configmap-configs could help mitigate


Based on these, I tend to lean towards number 3, unless of course we could actually get away with number 4 - but I don't feel like we can.